### PR TITLE
Write the k-modes chapter

### PR DESCRIPTION
**Pull request by daffidwilde**
Adapt the paper into a chapter.

### CODE REVIEW

File: chapters/kmodes/main.tex [hunk 1]
```diff
@@ -738,125 +785,127 @@ \subsection{Using number of classes for \(k\)}\label{subsec:nclasses}
...
+Here, \(C_{\mathrm{cao}}\) and \(C_{\mathrm{match}}\) are the final costs when a
 dataset \(\mathcal X\) is clustered using Cao's method and the proposed matching
-method respectively with \(k = 3\). For the sake of computational time, the
-proposed initialisation was given 25 repetitions as opposed to the 250
-repetitions in the remainder of this section. Apart from the sign of \(f\), the
-dataset generation processes used identical parameters in each case and the
-datasets considered here are all of comparable shape.
+method respectively with \(k = 3\). Given that costs should be minimised, this
+function is given to EDO to be minimised. As such, the case that uses \(f\) will
+be referred to as Cao-preferable. For the sake of computational time, the
```
Comment (drvinceknight) [inline chapters/kmodes/main.tex:830]: Is this worth having?

```suggestion
function is given to EDO to be minimised. As such, the case that uses \(f\) will
be referred to as Cao-preferable and \(-f\) will be referred to as matching-preferable. For the sake of computational time, the
```

File: chapters/kmodes/main.tex [hunk 2]
```diff
@@ -738,125 +785,127 @@ \subsection{Using number of classes for \(k\)}\label{subsec:nclasses}
...
     \centering
     \includegraphics[width=\imgwidth]{Fig9.pdf}
-    \caption{Histograms of fitness for the top performing percentile in each
-             case.}\label{fig:fitness}
+    \caption{%
+        Histograms of fitness for the top performing percentile in each case.
+    }\label{fig:fitness}
 \end{figure}
 
 This process yielded approximately 35,000 unique datasets for each case, and the
```
Comment (drvinceknight) [inline chapters/kmodes/main.tex:849]: Can we add a link to the archived data here?

File: chapters/kmodes/main.tex [hunk 3]
```diff
@@ -738,125 +785,127 @@ \subsection{Using number of classes for \(k\)}\label{subsec:nclasses}
...
+quantiles (the interquartile range, and lower and upper deciles).
 
-Figures~\ref{fig:edo_moments}~\&~\ref{fig:edo_quantiles} show the distributions
+Figures~\ref{fig:edo_moments}~and~\ref{fig:edo_quantiles} show the distributions
 of the six univariate statistics across all of the principal components in each
 case. In addition to this, they show a fitted Gaussian kernel density
 estimate~\cite{Bashtannyk2001} to accentuate the general shape of the
-histograms. What becomes immediately clear from each of these plots is that for
+histograms. What is immediately clear from each of these plots is that for
 datasets where Cao's method succeeds, the general spread of their first
```
Comment (drvinceknight) [inline chapters/kmodes/main.tex:885]:
```suggestion
datasets where Cao's method performs better, the general spread of their first
```

File: chapters/kmodes/main.tex [hunk 4]
```diff
@@ -905,24 +954,27 @@ \subsection{Artificial datasets}\label{subsec:artificial}
...
+described in Chapter~\ref{chp:edo} was utilised to extend this analysis to
+artificial datasets.
+
+The first of these analyses revealed that the proposed initialisation was able
+to consistently outperform both of the other methods when the choice of \(k\)
+had been optimised somewhat. However, the proposed method was unable to beat
+Cao's method when an external, and frankly separate, framework was imposed on
+each dataset by choosing \(k\) to be the number of classes present.
+
+The latter analysis showed that proposed method should be employed over Cao's if
```
Comment (drvinceknight) [inline chapters/kmodes/main.tex:977]:
```suggestion
The latter analysis showed that the proposed method should be employed over Cao's if
```

**Review by drvinceknight — CHANGES_REQUESTED** [review]
Looks great @daffidwilde, some of the suggestions are very much just suggestions. Let me know if you want to talk about any of it.

File: chapters/kmodes/main.tex [hunk 5]
```diff
@@ -121,20 +164,22 @@ \subsection{The \(k\)-modes algorithm}\label{subsec:kmodes}
...
+
+With the definition of a categorical distance metric, the notion of a
+representative point of a cluster can be addressed. When considering numeric
+data with \(k\)-means, a representative point is the mean of the points within
+the cluster. With categorical data, however, the mode is used as the measure for
+central tendency. This change follows from the concept of dissimilarity in that
+the point that best represents (i.e.\ is closest to) those in a cluster is one
+with the most frequent attribute values of the points in the cluster. The
+following definitions formalise this notion, and Theorem~\ref{thm:mode} provides
+a method to find such a point.
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex:177]: I think an example would be nice in this section but I'm not entirely sure what you mean.

File: chapters/kmodes/main.tex [hunk 6]
```diff
@@ -105,8 +134,22 @@ \subsection{The \(k\)-modes algorithm}\label{subsec:kmodes}
...
+distance measure to traditional Euclidean distance since it breaks down in a
+categorical attribute space; if the space were ordinal, like the natural
+numbers, then some sense of Euclidean distance could be recovered. However, in
+the absence of direction or what comes between two points, this distance has no
+sensible meaning.
+
+There are numerous measures available for the handling of categorical data.
+Definition~\ref{def:dissim} describes what is perhaps the simplest of those
+measures, and is the measure used in the seminal works by Huang. The pitfall of
+this measure is that it democratises the attribute space. In doing so, it does
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex:147]: Thank you very much, Vince 😄 I was a little worried about that one, I won't lie.

File: chapters/kmodes/main.tex [hunk 7]
```diff
@@ -194,170 +239,141 @@ \subsection{The \(k\)-modes algorithm}\label{subsec:kmodes}
...
-sets of players such that no pair of players is (rationally) unhappy with their
-matching. Algorithms to `solve' --- i.e.\ find stable matchings to --- instances
-of matching games are often structured to be party-oriented and aim to maximise
-some form of social or party-based optimality~\cite{%
-    Erdil2017,Fuku2006,Gale1962,Iwama2016,Kwanashie2015,Manlove2002%
-}.
+\subsection{The proposed method}\label{subsec:method}
+
+Both of the initialisation methods described in this section have a greedy
+component. Cao's method essentially chooses the densest point that has not
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex:345]: I believe it is but I prefer your wording.

File: chapters/kmodes/main.tex (hunk 2 above)
Comment (daffidwilde) [reply chapters/kmodes/main.tex:852]: Fixed by 80e2164

File: chapters/kmodes/main.tex [hunk 8]
```diff
@@ -105,8 +134,22 @@ \subsection{The \(k\)-modes algorithm}\label{subsec:kmodes}
...
+As has already been mentioned, the \(k\)-modes algorithm relies on another
+distance measure to traditional Euclidean distance since it breaks down in a
+categorical attribute space; if the space were ordinal, like the natural
+numbers, then some sense of Euclidean distance could be recovered. However, in
+the absence of direction or what comes between two points, this distance has no
+sensible meaning.
+
+There are numerous measures available for the handling of categorical data.
+Definition~\ref{def:dissim} describes what is perhaps the simplest of those
+measures, and is the measure used in the seminal works by Huang. The pitfall of
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex]: Fixed by 80e2164

File: chapters/kmodes/main.tex [hunk 9]
```diff
@@ -38,52 +39,80 @@ \chapter{A game-theoretic initialisation for the \(k\)-modes algorithm}
...
+\(k\)-modes algorithm makes use of similar principles to \(k\)-means, i.e.
+adjusting the partition toward some Voronoi tessellation according to a measure
+of centrality. However, the method allows for the clustering of categorical data
+by considering a cluster's modal values rather than its mean.
+
+The focus of this chapter is on the initialisation of the \(k\)-modes algorithm.
+Like \(k\)-means, there is no guarantee that any two runs of \(k\)-modes will
+provide the same clusters; this is due to the stochastic nature of its
+initialisation. Therefore, the performance of the algorithm is contingent upon
+the quality of this initial solution.
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex]: Fixed by 80e2164

File: chapters/kmodes/main.tex [hunk 10]
```diff
@@ -38,52 +39,80 @@ \chapter{A game-theoretic initialisation for the \(k\)-modes algorithm}
...
+    \item Section~\ref{sec:summary} concludes the chapter.
 \end{itemize}
 
 
 \subsection{The \(k\)-modes algorithm}\label{subsec:kmodes}
 
-The following notation will be used throughout this work to describe the objects
+Although the case study in Chapter~\ref{chp:edo} made use of the \(k\)-means
+algorithm, its notation was never formally introduced. Since this chapter
+considers the \(k\)-modes algorithm more directly, its notation is. As such, the
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex:106]: Yeah, I agree. I'll get rid of the waffle now 👍

[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex]: Fixed by be524c4

File: chapters/kmodes/main.tex [hunk 11]
```diff
@@ -38,52 +39,80 @@ \chapter{A game-theoretic initialisation for the \(k\)-modes algorithm}
...
+
+This chapter concludes with an analysis of all three initialisations and
+demonstrates that the proposed method can outperform both of the established
+initialisations using the traditional approach with benchmark datasets. In
+addition to this, the method introduced in Chapter~\ref{chp:edo} provides a more
+in-depth examination of the methods and how they perform against one another.
+The first stage of this analysis serves the purpose of highlighting the merits
+of each method in a familiar setting, and while the second stage bolsters these
+observations, it provides a vigorous defence of the proposed method and exposes
+the scenarios in which it excels.
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex]: Fixed by 0c7a869

File: chapters/kmodes/main.tex [hunk 12]
```diff
@@ -38,52 +39,80 @@ \chapter{A game-theoretic initialisation for the \(k\)-modes algorithm}
...
-The paper is structured as follows:
+The case study at the end of Chapter~\ref{chp:edo} examined the \(k\)-means
+algorithm --- a method for clustering a set of numeric data into \(k\) parts
+using Euclidean distance and the cluster means. This chapter considers another
+algorithm in the \(k\)-means paradigm, \(k\)-modes, described in a set of
+seminal papers by Huang~\cite{Huang1997a, Huang1997b, Huang1998}. The
+\(k\)-modes algorithm makes use of similar principles to \(k\)-means, i.e.
+adjusting the partition toward some Voronoi tessellation according to a measure
+of centrality. However, the method allows for the clustering of categorical data
+by considering a cluster's modal values rather than its mean.
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex:58]: Fixed by 0c7a869

File: chapters/kmodes/main.tex [hunk 13]
```diff
@@ -121,20 +164,22 @@ \subsection{The \(k\)-modes algorithm}\label{subsec:kmodes}
...
-
-With this metric, the notion of a representative point of a cluster is
-addressed. With numeric data and \(k\)-means, such a point is taken to be the
-mean of the points within the cluster. With categorical data, however, the mode
-is used as the measure for central tendency. This follows from the concept of
-dissimilarity in that the point that best represents (i.e.\ is closest to) those
-in a cluster is one with the most frequent attribute values of the points in the
-cluster. The following definitions and theorem formalise this and a method to
-find such a point.
+% TODO include proof of being distance metric?
```
[the remark this answers is not in the document]
Comment (daffidwilde) [reply chapters/kmodes/main.tex]: Fixed by 38ac5d3

**Pull request merged**